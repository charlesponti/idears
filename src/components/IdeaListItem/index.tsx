import t from 'prop-types'
import React, { useContext } from 'react'
import Button from 'src/components/Button'
import ListItem from 'src/components/ListItem'
import { addVoteToIdea } from 'src/context/Firebase'
import { Idea } from 'src/interfaces/Idea'
import { IdeasContext } from 'src/providers/IdeasProvider'
import { UserContext } from 'src/providers/UserProvider'
import { actionTypes } from 'src/reducers/action-types'
import styles from './IdeaListItem.module.scss'

interface IdeaListItemProps {
  idea: Idea
}

IdeaListItem.propTypes = {
  ideas: t.arrayOf(t.shape({
    _id: t.string,
    title: t.string,
    description: t.string,
    upvotes: t.number,
    downvotes: t.number
  }))
}

function IdeaListItem ({ idea: { id, title, description, upvotes, downvotes, ...idea } }: IdeaListItemProps) {
  const { dispatch } = useContext(IdeasContext)
  const { user } = useContext(UserContext)

  if (!user) return null

  async function upvoteIdea (_id: string) {
    dispatch({ type: actionTypes.UPVOTE_IDEA })
    await addVoteToIdea(_id, 1)
    dispatch({ type: actionTypes.IDEA_UPDATED })
  }

  async function downvoteIdea (_id: string) {
    dispatch({ type: actionTypes.DOWNVOTE_IDEA })
    await addVoteToIdea(_id, -1)
    dispatch({ type: actionTypes.IDEA_UPDATED })
  }

  const isUser = idea.user === user.uid
  const hasVoted = user.votes.indexOf(id) !== -1

  return (
    <ListItem>
      <div className={styles.title}>
        <p>{title}</p>
        <div className={styles.votes}>
          <p>
            upvotes: {upvotes}
          </p>
          <p>
            downvotes: {downvotes}
          </p>
        </div>
      </div>
      <div className={styles.description}>{description}</div>
      <div className={styles.btnContainer}>
        <Button
          aria-label="upvote idea"
          variant="success"
          onClick={() => upvoteIdea(id)}
          disabled={hasVoted || isUser}
        > Upvote </Button>
        <Button
          aria-label="downvote idea"
          variant="danger"
          onClick={() => downvoteIdea(id)}
          disabled={hasVoted || isUser}> Downvote </Button>
      </div>
    </ListItem>
  )
}

export default IdeaListItem
