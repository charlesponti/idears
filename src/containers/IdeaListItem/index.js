import React, { useContext } from 'react'
import styles from './IdeaListItem.module.scss'
import ListItem from '../../components/ListItem'
import { GlobalContext } from 'src/context/GlobalState'
import buttonStyles from '../../styles/button.module.scss'

function IdeaListItem({ idea: { _id, title, description, votes, ...idea  }}) {
  const { upvoteIdea, downvoteIdea, user } = useContext(GlobalContext)
  const isUser = user && idea.user === user._id
  return (
    <ListItem key={_id}> 
      <div className={styles.title}>
        <p>{title}</p>
        <p className={styles.votes}>
          Votes: {votes}
        </p>
      </div>
      <div className={styles.description}>{description}</div>
      <div className={styles.btnContainer}>
        <button
          aria-label="upvote idea"
          className={buttonStyles.btnSuccess} 
          onClick={() => upvoteIdea(_id)}
          disabled={isUser}
          > Upvote </button>
        <button 
          aria-label="downvote idea"
          className={buttonStyles.btnDanger}
          onClick={() => downvoteIdea(_id)} 
          disabled={isUser || votes === 0}> Downvote </button>
      </div>
    </ListItem>
  )
}

export default IdeaListItem
