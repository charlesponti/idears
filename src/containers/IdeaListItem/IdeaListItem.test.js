
import React from 'react'
import { render } from '@testing-library/react'
import IdeaListItem from '.'

describe('<IdeaListItem/>', () => {
  it('should render component', () => {
    const idea = { _id: 0, title: 'Some Idea', description: 'some description', votes: 0 }
    const { container } = render(<IdeaListItem idea={idea}/>)
    expect(container).toMatchSnapshot()
  })
})
