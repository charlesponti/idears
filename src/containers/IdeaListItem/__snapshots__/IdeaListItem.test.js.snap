// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IdeaListItem/> should render component 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="title"
    >
      <p>
        Some Idea
      </p>
      <p
        class="votes"
      >
        Votes: 
        0
      </p>
    </div>
    <div
      class="description"
    >
      some description
    </div>
    <div
      class="btnContainer"
    >
      <button
        class="btnSuccess"
      >
         Upvote 
      </button>
      <button
        class="btnDanger"
        disabled=""
      >
         Downvote 
      </button>
    </div>
  </div>
</div>
`;
